packages/i18n/src/index.ts

```diff
@@ -157,10 +158,135 @@ export const createI18nContext = (
 
 export const I18nContext = createContext<I18nContextInterface>({} as I18nContextInterface);
 export const useI18n = () => useContext(I18nContext);
+
+// -------------------------- Chained I18n -------------------
+export type I18nFormatOptions = Record<string, string | number>;
+
+interface I18nObject {
+  readonly [x: string]: string | ((...args: any) => any) | I18nObject;
+}
+
+type I18nFormatArgs = Record<string, string | number>;
+
+export type I18nPath<T extends I18nObject> = {
+  [K in keyof T]: T[K] extends I18nObject
+    ? I18nPath<T[K]>
+    : T[K] extends (options: infer OptionsArgs) => string
+    ? (options: OptionsArgs) => string
+    : (options?: I18nFormatArgs) => string;
+};
+
+export type Dictionaries<T extends I18nObject> = {
+  [key: string]: T;
+};
+
+type AddFunctionLengths<T> = T extends string
+  ? string
+  : T extends (...args: infer A) => unknown
+  ? T & { __length: A["length"] }
+  : {
+      [K in keyof T]: AddFunctionLengths<T[K]>;
+    };
+
+type RemoveFunctionLengths<T> = T extends string
+  ? string
+  : T extends (args: infer A) => infer R
+  ? (args: A) => R
+  : {
+      [K in keyof T]: RemoveFunctionLengths<T[K]>;
+    };
+
+type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (k: infer I) => void
+  ? I
+  : never;
+
+type GuaranteeIdenticalSignatures<T extends Dictionaries<I18nObject>> = Record<
+  keyof T,
+  RemoveFunctionLengths<UnionToIntersection<AddFunctionLengths<T[keyof T]>>>
+>;
+
+const buildI18nChain = <T extends I18nObject>(obj: T): I18nPath<T> => {
+  const keys = Object.keys(obj) as (keyof T)[];
+  const paths = keys.reduce((acc, key) => {
+    const value = obj[key] as any;
+    if (typeof value === "object") {
+      return {
+        ...acc,
+        [key]: buildI18nChain(value)
+      } as any;
+    } else if (typeof value === "function") {
+      return {
+        ...acc,
+        [key]: obj[key]
+      };
+    }
+    if (typeof value === "string") {
+      return {
+        ...acc,
+        [key]: (args: I18nFormatOptions) => {
+          return value.replace(/{{(.*?)}}/g, (_, key) => deepReadObject(args, key, ""));
+        }
+      };
+    } else {
+      throw new TypeError(
+        `Unsupported data format on the keys. Values must resolve to a string or a function that returns a string. Key name: "${
+          key as string
+        }"`
+      );
+    }
+  }, {} as Partial<I18nPath<T>>);
+
+  return paths as I18nPath<T>;
+};
+
+export interface I18nDictionary {
+  readonly [x: string]: string | ((...args: any) => string) | I18nDictionary;
+}
+
+export const makeChainedI18nContext = <
+  T extends Dictionaries<I18nObject>,
+  K extends keyof T
+>(props: {
+  dictionaries: T & GuaranteeIdenticalSignatures<T>;
+  locale: keyof T;
+  setContext?: boolean;
+}) => {
+  const [locale, _setLocale] = createSignal<K>(props.locale as K);
+  const [translate, _setTranslate] = createStore(buildI18nChain(props.dictionaries[props.locale]));
```

Comment:
Translations don't have to be a store either.
This is what meant in the previous comment.
All the properties of the dictionary objects will be mostly unique. Even if two languages will share the same word for something - there is no point in optimizing for that. You pretty much always want to rerun everything that is listening to the translations once the locale changes.
But since `t.hello` is nicer than `t().hello` you can do something like this:
https://playground.solidjs.com/anonymous/d7ff4242-94f3-43e5-85bc-f99dff3e6519

Reply:
I'll play around with this. Very interesting, never seen this before. Will be fun to learn

Reply:
Instead of doing it this way, I made the t function a proxy that returns the dictionaries with the current locale path.